### DISCUSSION TITLE
Error Message Improvements

### DISCUSSION THREAD

**Discussion by LandonSchropp** [mobxjs/mobx#2900] [Ideas]
Currently, when an error is thrown from an observer HOC, it produces an error message that looks like this:

```
Error: wrappedComponent(...): Nothing was returned from render. This usually means a return statement is missing. Or, to render nothing, return null.
```

This is very difficult to debug. *Which* component didn't return anything from render?

The stack traces also don't contain the name of the component. Here's an example from the  error above:

```
    at reconcileChildFibers (react-dom.development.js:14169)
    at reconcileChildren (react-dom.development.js:16990)
    at updateFunctionComponent (react-dom.development.js:17378)
    at updateSimpleMemoComponent (react-dom.development.js:17215)
    at updateMemoComponent (react-dom.development.js:17092)
    at beginWork (react-dom.development.js:19135)
    at HTMLUnknownElement.callCallback (react-dom.development.js:3945)
    at Object.invokeGuardedCallbackDev (react-dom.development.js:3994)
    at invokeGuardedCallback (react-dom.development.js:4056)
    at beginWork$1 (react-dom.development.js:23964)
    at performUnitOfWork (react-dom.development.js:22776)
    at workLoopSync (react-dom.development.js:22707)
    at renderRootSync (react-dom.development.js:22670)
    at performSyncWorkOnRoot (react-dom.development.js:22293)
    at scheduleUpdateOnFiber (react-dom.development.js:21881)
    at updateContainer (react-dom.development.js:25482)
    at react-dom.development.js:26021
    at unbatchedUpdates (react-dom.development.js:22431)
    at legacyRenderSubtreeIntoContainer (react-dom.development.js:26020)
    at Object.render (react-dom.development.js:26103)
    at Module.<anonymous> (index.jsx:9)
    at Module../source/javascript/index.jsx (index.jsx:11)
    at __webpack_require__ (bootstrap:851)
    at fn (bootstrap:150)
    at Object.1 (index.sass?ecf3:82)
    at __webpack_require__ (bootstrap:851)
    at checkDeferredModules (bootstrap:45)
    at Array.webpackJsonpCallback [as push] (bootstrap:32)
    at main.chunk.js:1
```

Given the way `observer` works, the only solution I can think of would be to pass an optional name parameter as the first argument.

```
export const Example = observer("Example", () => {
  ...
}
```

**Comment by urugator** [comment]
Have not investigated, but possibly related to https://github.com/mobxjs/mobx/issues/2721

**Reply by LandonSchropp** [reply]
I didn't even realize there was a `displayName` property. 🙂

**Reply by iChenLei** [reply]
Haha, it's React component usage. https://reactjs.org/docs/higher-order-components.html#convention-wrap-the-display-name-for-easy-debugging

**Reply by LandonSchropp** [reply]
Ah, that's why I couldn't find it. Thanks for the tip!

**Reply by LandonSchropp** [reply]
This didn't really work for me. All of my components now have a `displayName` property, but I still get indecipherable error messages:

<img width="814" alt="Screen Shot 2021-04-27 at 4 28 45 PM" src="https://user-images.githubusercontent.com/361591/116315137-b5eaac80-a775-11eb-9e75-495976b39e32.png">

It's very hard to debug the root cause with a stack trace like this.

**Reply by iChenLei** [reply]
If you use `observer` from `mobx-react` or `mobx-react-lite`, we will create a react memo component to wrap your component internally. But React memo component can't set displayName multi times(<React 17, even can't set displayName). It's upstream problem so we can only wait react official to fix it . look for more details -> https://github.com/facebook/react/pull/20659.

**Reply by urugator** [reply]
You can either use named function `observer(function MyNamedComponent() {})` or you can set `displayName` before wrapping function in `observer`

**Reply by LandonSchropp** [reply]
Is there a way MobX could be updated so you wouldn't have to set the `displayName` on the inner function? Could `set displayName` on the HOC set the inner `displayName`?

**Reply by LandonSchropp** [reply]
Thanks for all the replies by the way!

**Reply by LandonSchropp** [reply]
Using the named function doesn't work if you want to export a component with the same name.

``` jsx
function Example() {
  return <p>Hello</p>;
}

// Doesn't work!
export const Example = observer(Example);
```

The second approach works but feels very clunky.

``` jsx
const ExampleInner = () => {
  return <p>Hello</p>;
};

ExampleInner.displayName = "Example";

export const Example = observer(ExampleInner);
```

A property setter would be much easier and cleaner, and would feel like setting `propTypes`.

``` jsx
export const Example = observer(() => {
  return <p>Hello</p>;
});

Example.displayName = "Example";
```

**Reply by urugator** [reply]
It's PITA, but if you use `eslint-plugin-react` you will face exactly the same problem with `React.memo`.
```javascript
export const Example = observer(
  function Example() {
    return <p>Hello</p>;
  }
)
```

**Reply by LandonSchropp** [reply]
I dug through the code a bit and I think I've caught up to where the conversation is at. Let me summarize to see if I understand.

The memoized component *should* be getting its `displayName` set in the code [here](https://github.com/mobxjs/mobx/blob/72d06f8cd2519ce4dbfb807bc13556ca35866690/packages/mobx-react-lite/src/observer.ts#L76). However, there's a bug in React that's preventing this from working. @iChenLei opened facebook/react#20659 to address the core issue, so hopefully they'll accept that PR soon and this will get fixed.

Does that sound right? Sorry for the slowness on my part here.

**Reply by iChenLei** [reply]
https://github.com/facebook/react/pull/21392 has merged, so maybe react 17.0.3 will solve this problem.

**Comment by mweststrate** [comment]
I'm not sure if that is possible since React creates that property already
as well IIRC, but feel free to give it a try hiding it with our own, might
be a nice work around for now indeed :)

**Reply by LandonSchropp** [reply]
I'll poke around in the code and see what I can come up with.
